File: backend/api/projects.py
"""
CRUD API for Projects
"""
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from pydantic import BaseModel
from typing import Optional

from database import get_db
from models.db_models import User, Project
from api.auth import get_current_user

router = APIRouter()


class ProjectCreate(BaseModel):
    name: str
    description: Optional[str] = ""


class ProjectUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None


class ProjectResponse(BaseModel):
    id: int
    name: str
    description: str
    scan_count: int = 0
    created_at: str
    updated_at: str

    class Config:
        from_attributes = True


def _project_to_response(p: Project) -> ProjectResponse:
    return ProjectResponse(
        id=p.id,
        name=p.name,
        description=p.description or "",
        scan_count=len(p.scans),
        created_at=p.created_at.isoformat(),
        updated_at=p.updated_at.isoformat(),
    )


@router.get("", response_model=list[ProjectResponse])
async def list_projects(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """List all projects for the current user."""
    projects = db.query(Project).filter(Project.user_id == current_user.id).order_by(Project.updated_at.desc()).all()
    return [_project_to_response(p) for p in projects]


@router.post("", response_model=ProjectResponse)
async def create_project(
    body: ProjectCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Create a new project."""
    project = Project(
        user_id=current_user.id,
        name=body.name,
        description=body.description or "",
    )
    db.add(project)
    db.commit()
    db.refresh(project)
    return _project_to_response(project)


@router.get("/{project_id}", response_model=ProjectResponse)
async def get_project(
    project_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Get a project by ID."""
    project = db.query(Project).filter(Project.id == project_id, Project.user_id == current_user.id).first()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    return _project_to_response(project)


@router.put("/{project_id}", response_model=ProjectResponse)
async def update_project(
    project_id: int,
    body: ProjectUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Update a project."""
    project = db.query(Project).filter(Project.id == project_id, Project.user_id == current_user.id).first()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    if body.name is not None:
        project.name = body.name
    if body.description is not None:
        project.description = body.description
    db.commit()
    db.refresh(project)
    return _project_to_response(project)


@router.delete("/{project_id}")
async def delete_project(
    project_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Delete a project and all its scans."""
    project = db.query(Project).filter(Project.id == project_id, Project.user_id == current_user.id).first()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    db.delete(project)
    db.commit()
    return {"ok": True}

File: frontend/src/api/projects.ts
import axios from 'axios';
import { getApiBaseUrl } from '../lib/apiBase';

const API_BASE_URL = getApiBaseUrl();

const getAuthHeaders = () => {
  const token = localStorage.getItem('token');
  return token ? { Authorization: `Bearer ${token}` } : {};
};

export interface Project {
  id: number;
  name: string;
  description: string;
  scan_count: number;
  created_at: string;
  updated_at: string;
}

export const listProjects = async (): Promise<Project[]> => {
  const response = await axios.get(`${API_BASE_URL}/api/projects`, {
    headers: getAuthHeaders(),
  });
  return response.data;
};

export const createProject = async (name: string, description?: string): Promise<Project> => {
  const response = await axios.post(
    `${API_BASE_URL}/api/projects`,
    { name, description: description || '' },
    { headers: getAuthHeaders() }
  );
  return response.data;
};

export const getProject = async (id: number): Promise<Project> => {
  const response = await axios.get(`${API_BASE_URL}/api/projects/${id}`, {
    headers: getAuthHeaders(),
  });
  return response.data;
};

export const updateProject = async (id: number, data: { name?: string; description?: string }): Promise<Project> => {
  const response = await axios.put(`${API_BASE_URL}/api/projects/${id}`, data, {
    headers: getAuthHeaders(),
  });
  return response.data;
};

export const deleteProject = async (id: number): Promise<void> => {
  await axios.delete(`${API_BASE_URL}/api/projects/${id}`, {
    headers: getAuthHeaders(),
  });
};
